### PR TITLE
feat(code): switch to typescript, overhaul test, fix bugs

### PR DESCRIPTION
**Pull request by blaskovicz**

### CODE REVIEW

**Comment by blaskovicz** [comment]
This is a big, opinionated change, but I think typescript is great for maintainability and refactoring long term. Let me know if you have any questions.  At the end of this change, the tests now test more (while stubbing requests) so CI can easily be set up. Build products still output to the distribution directory in es5 format with node-style imports.

**Comment by ColinMcNeil** [comment]
I think this is probably a good idea. The whole point of this vs. tplink-smarthome-api is that this is much more minimal and better to compile for browsers. Switching to TS and removing dependencies is a good step.

File: lib/api-utils.ts [hunk 1]
```diff
@@ -0,0 +1,7 @@
+export function checkError(response) {
+  if (response && response.data && response.data.error_code !== 0) {
```
Comment (ColinMcNeil) [inline lib/api-utils.ts]: We could split this out into multiple branches...that way if there's no response a different error occurs

**Comment by ColinMcNeil** [comment]
You should add https://palantir.github.io/tslint/

File: lib/api-utils.ts (hunk 1 above)
Comment (blaskovicz) [reply lib/api-utils.ts]: probably not a bad idea. i also saw on the tplink website in some of the camera code that there is some other strange logic they used. Not sure if we need it here, but:

https://d2baw28zdve8g3.cloudfront.net/ssl/tpcamera-admin/js/ipc-business-libs.js

```js
                var errCodeStrIndex = inputArgs["errCodeStrIndex"] || "errorCode";
                var noErrorCode = inputArgs["noErrorCode"] || 0;
                var defaultErrorCode = inputArgs["defaultErrorCode"] || -1;

                if (response[errCodeStrIndex] == noErrorCode) {
...

**Comment by blaskovicz** [comment]
I can add tslint. I think I just used the default tslint and prettier settings. I can also write a .prettierrc

**Comment by blaskovicz** [comment]
tslint:latest or tslint:recommended? I'm also adding [tslint-config-prettier](https://alexjoverm.github.io/2017/06/12/Use-Prettier-with-TSLint-and-be-happy/).

**Comment by blaskovicz** [comment]
tslint pull opened in #21

**Pull request merged**